File: app/studio-backend/src/studio_service/schemas/canvas.py
"""OpenAPI models for S2 canvas endpoints (React Flow consumer)."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

ValidationStatus = Literal["pass", "warn", "fail", "not_run"]


class SourceRef(BaseModel):
    ref_type: str
    ref: str
    summary: str | None = None


class ValidationOverlaySummary(BaseModel):
    id: str
    status: ValidationStatus
    check_type: str
    message_count: int
    source_refs: list[SourceRef]


class CanvasNode(BaseModel):
    id: str
    graph_node_id: str
    label: str
    type: str
    category: str
    source_refs: list[SourceRef]
    annotations: list[dict[str, str]] = Field(default_factory=list)
    validation_overlays: list[ValidationOverlaySummary] = Field(default_factory=list)


class CanvasEdge(BaseModel):
    id: str
    graph_edge_id: str
    source: str
    target: str
    relation: str
    source_refs: list[SourceRef]
    validation_overlays: list[ValidationOverlaySummary] = Field(default_factory=list)
    label: str | None = None
    summary: str | None = None
    agent: str | None = None
    skill: str | None = None


class CanvasOverlay(BaseModel):
    id: str
    validation_ref: str
    target: dict[str, str]
    check_type: str
    status: ValidationStatus
    messages: list[dict[str, Any]]
    source_refs: list[SourceRef]


class CanvasSection(BaseModel):
    id: str
    report_ref: str
    kind: str
    title: str
    status: str
    summary: str
    items: list[dict[str, Any]]
    source_refs: list[SourceRef]


class CanvasLegendCategory(BaseModel):
    id: str
    label: str


class CanvasLegend(BaseModel):
    categories: list[CanvasLegendCategory]
    validation_statuses: dict[str, int]


class CanvasFilters(BaseModel):
    section: str | None = None
    validation_status: ValidationStatus | None = None
    entity_type: str | None = None
    q: str | None = None


class CanvasMeta(BaseModel):
    total_nodes: int
    filtered_nodes: int
    total_edges: int
    filtered_edges: int


class CanvasFullResponse(BaseModel):
    """Stable S2 payload for React Flow mapViewModel.ts."""

    canvas: dict[str, Any]
    nodes: list[CanvasNode]
    edges: list[CanvasEdge]
    overlays: list[CanvasOverlay]
    sections: list[CanvasSection]
    legend: CanvasLegend
    filters: CanvasFilters
    meta: CanvasMeta


class CanvasNodeDetailResponse(BaseModel):
    node: CanvasNode
    overlays: list[CanvasOverlay]

File: app/studio-frontend/src/types/canvas.ts
export type ValidationStatus = "pass" | "warn" | "fail" | "not_run";

export type SourceRef = {
  ref_type: string;
  ref: string;
  summary?: string | null;
};

export type ValidationOverlaySummary = {
  id: string;
  status: ValidationStatus;
  check_type: string;
  message_count: number;
  source_refs: SourceRef[];
};

export type CanvasNode = {
  id: string;
  graph_node_id: string;
  label: string;
  type: string;
  category: string;
  source_refs: SourceRef[];
  annotations?: { kind: string; text: string }[];
  validation_overlays: ValidationOverlaySummary[];
};

export type CanvasEdge = {
  id: string;
  graph_edge_id: string;
  source: string;
  target: string;
  relation: string;
  source_refs: SourceRef[];
  validation_overlays: ValidationOverlaySummary[];
  label?: string | null;
  summary?: string | null;
  /** Present on workflow-builder edges from transitions.yaml */
  agent?: string | null;
  skill?: string | null;
};

export type CanvasOverlay = {
  id: string;
  validation_ref: string;
  target: Record<string, string>;
  check_type: string;
  status: ValidationStatus;
  messages: Record<string, unknown>[];
  source_refs: SourceRef[];
};

export type CanvasLegendCategory = {
  id: string;
  label: string;
};

export type CanvasLegend = {
  categories: CanvasLegendCategory[];
  validation_statuses: Record<ValidationStatus, number>;
};

export type CanvasFilters = {
  section: string | null;
  validation_status: ValidationStatus | null;
  entity_type: string | null;
  q: string | null;
};

export type CanvasMeta = {
  total_nodes: number;
  filtered_nodes: number;
  total_edges: number;
  filtered_edges: number;
};

export type CanvasFullResponse = {
  canvas: {
    id: string;
    name?: string;
    authority?: string;
  };
  nodes: CanvasNode[];
  edges: CanvasEdge[];
  overlays: CanvasOverlay[];
  sections: Record<string, unknown>[];
  legend: CanvasLegend;
  filters: CanvasFilters;
  meta: CanvasMeta;
};

export type CanvasNodeDetailResponse = {
  node: CanvasNode;
  overlays: CanvasOverlay[];
};

export type CanvasFilterParams = {
  section?: string;
  validation_status?: ValidationStatus;
  entity_type?: string;
  q?: string;
};
